lang/java/avro/src/main/java/org/apache/avro/file/BZip2Codec.java

```diff
@@ -42,64 +40,20 @@ protected Codec createInstance() {
   public String getName() { return DataFileConstants.BZIP2_CODEC; }
 
   @Override
-  public ByteBuffer compress(ByteBuffer uncompressedData) throws IOException {
-
-    ByteArrayOutputStream baos = getOutputBuffer(uncompressedData.remaining());
-    BZip2CompressorOutputStream outputStream = new BZip2CompressorOutputStream(baos);
-
-    try {
-      outputStream.write(uncompressedData.array(),
-                         uncompressedData.position(),
-                         uncompressedData.remaining());
-    } finally {
-      outputStream.close();
-    }
-
-    ByteBuffer result = ByteBuffer.wrap(baos.toByteArray());
-    return result;
+  protected OutputStream compressedStream(OutputStream output)
+      throws IOException {
+    return new BZip2CompressorOutputStream(output);
   }
 
   @Override
-  public ByteBuffer decompress(ByteBuffer compressedData) throws IOException {
-    ByteArrayInputStream bais = new ByteArrayInputStream(compressedData.array());
-    BZip2CompressorInputStream inputStream = new BZip2CompressorInputStream(bais);
```

Comment:
Most of the Codecs are internally based on InputStreams and OutputStreams.  I refactored the commonalities out into two abstract classes (these would be better as 'mix-in' interfaces in Java 8+).